# Variables, operadores y expresiones

Al ser un lenguaje de alto nivel, Python dispone de los tipos de datos elementales en cualquier lenguaje de programación, 
pero además incluye estructuras de datos más complejas y con altas prestaciones que facilitan en muchos 
aspectos la tarea del programador.

Python es un lenguaje de tipado dinámico en el que no hace falta declarar el tipo de dato que asignará a una variable, 
de igual manera una variable puede cambiar de tipo conforme la ejecución del programa, por ello se debe tener cuidado 
con la sintaxis para definir cada tipo de dato.

## Variables

Cuando escribimos un programa necesitamos una forma de identificar los valores con los que vamos a trabajar. Por ejemplo, en un cálculo de ingeniería podríamos necesitar representar una masa, una velocidad o una temperatura. Para ello podemos utilizar variables.

En Python, una variable es un nombre que hace referencia a un objeto. La asociación entre un nombre y un objeto se realiza mediante una asignación, utilizando el operador `=`:

In [31]:
masa = 25
velocidad = 10.0
nombre = "Catalina"

En estas instrucciones, los nombres `masa`, `velocidad` y `nombre` quedan asociados con los valores `25`, `10.0` y `"Catalina"`, respectivamente.

Una forma útil de visualizar este proceso consiste en pensar que el nombre de una variable actúa como una etiqueta asociada a un objeto:

```mermaid
flowchart LR
    masa["masa"] --> valor_masa["25"]
    velocidad["velocidad"] --> valor_velocidad["10.0"]
    nombre["nombre"] --> valor_nombre["'Catalina'"]
```

En este sentido, la instrucción

```python
masa = 25
```

puede interpretarse como: asocia el nombre `masa` con el objeto que representa el valor `25`.

### Reasignación de variables

Una variable puede asociarse posteriormente con otro objeto mediante una nueva asignación:

In [43]:
masa = 25
masa = 30

Después de ejecutar la segunda instrucción, el nombre `masa` hace referencia al valor `30`.

Esto pone de manifiesto una diferencia importante entre la asignación en programación y la igualdad matemática. En Python, el símbolo `=` no establece una igualdad, sino que realiza una asignación.

Por ejemplo:

In [46]:
x = 5
x = x + 1

En matemáticas, una expresión como

$$
x = x + 1
$$

no puede cumplirse. Sin embargo, en Python la segunda instrucción tiene otro significado: primero se evalúa la expresión situada a la derecha, `x + 1`, y después el resultado se vuelve a asociar con el nombre `x`.

Así, después de ejecutar las instrucciones anteriores, `x` queda asociado con el valor `6`.

### Varios nombres y asignación múltiple

Considera ahora el siguiente ejemplo:

In [53]:
m = 7
n = m

Primero se asocia el nombre `m` con el valor `7`. Después, mediante la asignación `n = m`, el nombre `n` queda asociado con el mismo objeto al que hace referencia `m`.

```mermaid
flowchart LR
    m["m"] --> valor["7"]
    n["n"] --> valor
```

Por ahora basta con comprender que una variable es un nombre asociado con un objeto. Más adelante veremos que esta característica tiene consecuencias importantes cuando trabajamos con estructuras de datos y otros tipos de objetos.

Python también permite realizar varias asignaciones en una sola instrucción:

In [66]:
masa, velocidad = 25, 10

En este caso, `masa` queda asociado con `25` y `velocidad` con `10`. Para este caso, la instrucción anterior es equivalente a escribir:

In [69]:
masa = 25
velocidad = 10

### Reglas para nombrar variables

Los nombres de variables deben cumplir algunas reglas sintácticas.

Los nombres pueden contener letras, números y guiones bajos (`_`):

In [76]:
esfuerzo_axial = 150e6
diametro_01 = 30e-3
nombre = "Ximena"

Un nombre puede comenzar con una letra o con un guion bajo, pero no con un número. Por ejemplo:

In [80]:
temperatura = 35
deformacion_unitaria = 300e-6

son nombres válidos, mientras que:

In [84]:
30_presion = 30e3

SyntaxError: invalid decimal literal (298741684.py, line 1)

produce un error de sintaxis.

Los espacios tampoco están permitidos en los nombres de variables. Por ejemplo:

In [87]:
modulo de elasticidad = 70e9

SyntaxError: invalid syntax (3836078760.py, line 1)

no es válido. Si queremos utilizar un nombre formado por varias palabras, podemos separarlas mediante guiones bajos:

In [95]:
modulo_de_elasticidad = 70e9

Python también distingue entre letras mayúsculas y minúsculas. Por tanto:

In [91]:
nombre = "Ana"
Nombre = "Paola"
NOMBRE = "Ximena"

define tres variables diferentes.

Esta característica suele expresarse diciendo que Python es un lenguaje *case sensitive*.

### Palabras clave

Algunas palabras forman parte de la sintaxis del lenguaje y no pueden utilizarse como nombres de variables. Estas palabras se conocen como **palabras clave** o *keywords*.

Por ejemplo:

In [98]:
lambda = 1500

SyntaxError: invalid syntax (2106972384.py, line 1)

produce un `SyntaxError`, ya que `lambda` es una palabra clave de Python.

La lista de palabras clave disponibles en la versión de Python que se esté utilizando puede consultarse mediante el módulo `keyword`:

```python
import keyword

print(keyword.kwlist)
```

Entre ellas se encuentran palabras como `if`, `else`, `for`, `while`, `def`, `class`, `True`, `False` y `None`.

### Buenas prácticas para nombrar variables

Además de cumplir las reglas sintácticas, es recomendable elegir nombres que hagan que el código sea fácil de leer y comprender.

En general, conviene utilizar nombres **descriptivos**. Por ejemplo:

In [106]:
diametro = 0.03

suele ser más informativo que:

In [109]:
d = 0.03

aunque los nombres cortos pueden resultar adecuados cuando tienen un significado claro en el contexto, como ocurre frecuentemente con variables matemáticas.

En Python es habitual utilizar la convención denominada `snake_case`, que consiste en escribir las palabras en minúsculas y separarlas mediante guiones bajos:

In [113]:
velocidad_angular = 10
temperatura_ambiente = 25
modulo_de_elasticidad = 70e9

Esta convención facilita la lectura de nombres formados por varias palabras y es la recomendada habitualmente para variables y funciones en Python.

También conviene evitar utilizar nombres que Python ya proporciona para funciones o tipos incorporados. Por ejemplo:

In [118]:
sum = 10

es una instrucción válida, pero hace que el nombre `sum` deje de referirse a la función incorporada que permite sumar elementos.

Por esta razón, suele ser conveniente evitar nombres como:

```python
sum
print
list
int
float
type
```

cuando su uso pueda ocultar las funciones o tipos correspondientes.

Puedes encontrar recomendaciones adicionales sobre el estilo de los nombres en la guía oficial [**PEP 8 – Style Guide for Python Code**](https://peps.python.org/pep-0008/).

Hasta ahora hemos utilizado valores enteros, decimales y cadenas de caracteres sin detenernos en las diferencias entre ellos. En la siguiente sección estudiaremos los tipos de datos básicos disponibles en Python.

## Tipos de datos básicos

En la sección anterior vimos que una variable es un nombre asociado con un objeto. Dependiendo de la información que representen, los objetos pueden ser de distintos tipos: números enteros, números de punto flotante, valores lógicos, texto, entre otros.

El **tipo de dato** indica qué clase de valor representa un objeto y qué operaciones pueden realizarse con él.

Podemos consultar el tipo de un objeto mediante la función `type`:

In [139]:
masa = 25
temperatura = 24.5
nombre = "motor"

In [142]:
type(masa)

int

In [144]:
type(temperatura)

float

In [146]:
type(nombre)

str

En estos casos, Python identifica los tipos `int`, `float` y `str`, respectivamente.

A continuación veremos algunos de los tipos básicos que utilizaremos con frecuencia a lo largo de estos apuntes.

### Números enteros: `int`

Los números enteros se representan mediante el tipo `int`. Este tipo permite trabajar con valores positivos, negativos o iguales a cero, siempre que no tengan parte decimal.

Por ejemplo:

In [151]:
numero_dientes = 24
muestras = 1000
indice = -1

Podemos comprobar el tipo de uno de estos valores mediante `type`:

In [154]:
type(numero_dientes)

int

Los valores de tipo `int` son adecuados para representar cantidades discretas, como el número de elementos de un sistema, la cantidad de muestras adquiridas o el número de iteraciones de un procedimiento.

### Números de punto flotante: `float`

Los números que incluyen una parte decimal se representan habitualmente mediante el tipo `float`.

Por ejemplo:

In [158]:
temperatura = 23.5
longitud = 0.125
densidad = 7850.0

También es posible escribir números de punto flotante mediante **notación científica**:

In [162]:
modulo_de_elasticidad = 70e9
deformacion = 250e-6

Las expresiones anteriores corresponden a

$$
70\times 10^9
$$

y

$$
250\times 10^{-6},
$$

respectivamente.

Este tipo de notación resulta especialmente útil en ingeniería, donde es frecuente trabajar con magnitudes muy grandes o muy pequeñas. Conviene notar que un valor como `70e9` es de tipo `float`, aunque no incluya explícitamente un punto decimal.

In [170]:
type(modulo_de_elasticidad)

float

:::{note} `float` y la precisión
Los números de tipo `float` tienen una precisión finita. Esto significa que muchos números reales no pueden representarse exactamente en una computadora. Retomaremos esta cuestión más adelante cuando estudiemos expresiones numéricas.
:::

### Números complejos: `complex`

Python también permite trabajar directamente con números complejos mediante el tipo `complex`.

Un número complejo puede escribirse utilizando `j` para representar la unidad imaginaria:

In [182]:
z = 3 + 4j
type(z)

complex

En matemáticas es frecuente utilizar $i$ para representar la unidad imaginaria. En Python se utiliza `j`, una notación habitual también en áreas como la ingeniería eléctrica.

Por ejemplo, una impedancia eléctrica podría representarse mediante:

In [185]:
impedancia = 10 + 5j

Por ahora basta con reconocer la forma de escritura de este tipo de valores.

### Valores booleanos: `bool`

El tipo `bool` representa valores lógicos y sólo puede tomar dos valores:

```python
True
False
```

Por ejemplo:

In [191]:
motor_encendido = True
alarma_activa = False

In [193]:
type(motor_encendido)

bool

In [195]:
type(alarma_activa)

bool

Los valores booleanos serán especialmente importantes cuando trabajemos con comparaciones y condiciones. Más adelante veremos que expresiones como las comparaciones producen precisamente resultados de tipo `bool`.

Es importante escribir `True` y `False` con la primera letra en mayúscula, ya que estas palabras forman parte del lenguaje Python.

### Cadenas de caracteres: `str`

Python utiliza el tipo `str` para representar texto. Una cadena de caracteres puede escribirse delimitando su contenido mediante comillas simples o dobles:

In [207]:
material = "acero"
unidad = 'MPa'

Por el momento únicamente necesitamos reconocer `str` como uno de los tipos básicos disponibles en Python. En un capítulo posterior estudiaremos con mayor detalle cómo crear, manipular y procesar cadenas de caracteres.

### Una variable puede referirse a objetos de distintos tipos

Una variable no queda asociada permanentemente con un único tipo de dato.

Considera el siguiente ejemplo:

In [212]:
dato = 10
type(dato)

int

Inicialmente, `dato` hace referencia a un objeto de tipo `int`. Podemos realizar una nueva asignación:

In [217]:
dato = 10.5
type(dato)

float

Ahora el mismo nombre hace referencia a un objeto de tipo `float`.

También podríamos escribir:

In [220]:
dato = "temperatura"
type(dato)

str

y `dato` pasaría a hacer referencia a una cadena de caracteres.

Este comportamiento suele describirse diciendo que Python utiliza **tipado dinámico**: el nombre de una variable no queda fijado a un único tipo, sino que puede referirse a objetos de distintos tipos a lo largo de la ejecución de un programa.

En ocasiones necesitaremos transformar un valor de un tipo a otro. Por ejemplo, convertir un número entero en uno de punto flotante o transformar una cadena que contiene dígitos en un valor numérico. Este será el tema de la siguiente sección.

:::{note .dropdown} Un resumen de los tipos de datos
Los tipos presentados en esta sección pueden resumirse de la siguiente manera:

| Tipo | Ejemplo | Representa |
|---|---|---|
| `int` | `25` | números enteros |
| `float` | `3.14` | números de punto flotante |
| `complex` | `2 + 3j` | números complejos |
| `bool` | `True` | valores lógicos |
| `str` | `"motor"` | texto |

Estos no son los únicos tipos de datos disponibles en Python, pero son suficientes para comenzar a construir expresiones y resolver problemas sencillos.
:::

## Conversión de tipos

En algunas situaciones necesitamos transformar un valor de un tipo de dato a otro. Por ejemplo, podemos convertir un número entero en uno de punto flotante, transformar una cadena que contiene un número en un valor numérico o convertir un valor en texto.

Python proporciona varias funciones incorporadas para realizar estas conversiones de manera explícita:

| Función | Conversión |
|---|---|
| `int(x)` | convierte `x` a un número entero |
| `float(x)` | convierte `x` a un número de punto flotante |
| `complex(x)` | convierte `x` a un número complejo |
| `bool(x)` | convierte `x` a un valor booleano |
| `str(x)` | convierte `x` a una cadena de caracteres |

No todas las conversiones son posibles y, en algunos casos, una conversión puede provocar pérdida de información.

### Conversión a números enteros

La función `int` permite convertir valores compatibles al tipo `int`. Por ejemplo:

In [245]:
int(3.8)

3

Al convertir un número de punto flotante a entero, Python elimina la parte decimal:

In [249]:
int(7.9)

7

De manera similar:

In [256]:
int(-2.8)

-2

Por tanto, `int` no realiza un redondeo, sino que elimina la parte fraccionaria y conserva únicamente la parte entera.

También es posible convertir una cadena de caracteres si su contenido representa correctamente un número entero:

In [259]:
int("25")

25

En este caso, el resultado es el número entero `25`.

Sin embargo, una cadena como:

In [261]:
int("25.5")

ValueError: invalid literal for int() with base 10: '25.5'

no puede convertirse directamente a `int`, ya que `"25.5"` representa un número con parte decimal.

### Conversión a números de punto flotante

La función `float` permite convertir valores compatibles al tipo `float`.

Por ejemplo:

In [269]:
float(5)

5.0

In [271]:
float("3.14")

3.14

In [273]:
float("1.5e3")

1500.0


La conversión desde cadenas resulta especialmente útil cuando los datos numéricos provienen de texto, por ejemplo, de una entrada proporcionada por un usuario o de un archivo.

### Conversión a números complejos

La función `complex` permite crear números complejos a partir de valores compatibles. Por ejemplo:

In [281]:
complex(5)

(5+0j)

También podemos especificar directamente la parte real y la parte imaginaria:

In [284]:
complex(3, 4)

(3+4j)

O a partir de un string:

In [295]:
complex("-4+2j")

(-4+2j)

### Conversión a valores booleanos

La función `bool` permite convertir otros valores al tipo `bool`. Para valores numéricos, el comportamiento básico es sencillo:

In [304]:
bool(0)

False

In [306]:
bool(1)

True

In [308]:
bool(-3)

True

In [310]:
bool(2.5)

True

El número cero se convierte en `False`, mientras que cualquier valor numérico distinto de cero se convierte en `True`. Más adelante veremos que otros tipos de objetos también pueden interpretarse como verdaderos o falsos cuando se utilizan en condiciones.

### Conversión a cadenas de caracteres

La función `str` permite obtener una representación textual de un valor. Por ejemplo:

In [317]:
str(25)

'25'

In [319]:
str(3.14)

'3.14'

In [321]:
str(True)

'True'

Es importante distinguir entre un valor numérico y una cadena que contiene los mismos caracteres. Por ejemplo:

```python
25
```

y:

```python
"25"
```

no representan el mismo tipo de dato. Podemos comprobarlo mediante `type`:

In [328]:
type(25)

int

In [330]:
type("25")

str

El primero es un objeto de tipo `int`, mientras que el segundo es un objeto de tipo `str`.

### Conversiones que pueden perder información

Una conversión no siempre conserva toda la información del valor original. Por ejemplo:

In [347]:
temperatura = 23.8
temperatura_entera = int(temperatura)

Después de la conversión:

In [350]:
temperatura_entera

23

La parte decimal se ha perdido.

:::{warning}
Una conversión crea un valor de otro tipo y, dependiendo de los tipos involucrados, puede modificar o perder parte de la información original.
:::

Además, una conversión puede fallar si el valor no es compatible con el tipo de destino. Por ejemplo:

In [356]:
float("temperatura")

ValueError: could not convert string to float: 'temperatura'

produce un error porque la cadena `"temperatura"` no contiene una representación válida de un número.

### Operaciones entre distintos tipos numéricos

No siempre es necesario realizar una conversión explícita antes de combinar valores numéricos de distintos tipos.

Por ejemplo:

In [360]:
2 + 3.5

5.5

combina un valor de tipo `int` con uno de tipo `float`. El resultado es de tipo `float`:

In [364]:
type(2 + 3.5)

float

Cuando una operación combina ciertos tipos numéricos compatibles, Python determina automáticamente un tipo adecuado para representar el resultado.

Esto no significa que los objetos originales cambien de tipo. En el ejemplo anterior, `2` continúa siendo un valor de tipo `int`; simplemente el resultado de la operación se representa mediante un `float`.

Algo similar ocurre al combinar números reales con números complejos:

In [371]:
2 + 3j

(2+3j)

En este caso, el resultado es de tipo `complex`.

## Operadores aritméticos

En cualquier lenguaje de programación los operadores aritméticos permiten realizar las operaciones aritméticas básicas con tipos numéricos, además de que permiten formar expresiones compuestas. En la {numref}`op-ar` se muestran los operadores aritméticos disponibles en Python y la operación que realizan.

```{list-table} Operadores aritméticos
:header-rows: 1
:name: op-ar

* - Operación 
  - Operador
* - Suma
  - `+`
* - Resta
  - `-`
* - Multiplicación
  - `*` 
* - División
  - `/`
* - División entera
  - `//`
* - Módulo
  - `%`
* - Exponenciación
  - `**`
```

Enseguida se muestran operaciones de ejemplo con cada uno de los operadores aritméticos listados anteriormente:

In [4]:
1000 + 2000 # suma

3000

In [5]:
500 - 350 # resta

150

In [6]:
352 * 34 # multiplicación

11968

In [7]:
3/8 # división

0.375

In [8]:
10 // 3 # división entera

3

In [9]:
10 % 3 # módulo

1

In [10]:
10**3 # exponenciación

1000

## Operadores relacionales

Los operadores relacionales (o de comparación) nos permite efectuar comparaciones entre objetos de Python. El resultado de una comparación es un valor booleano `True` o `False`. A continuación se ejemplifican los operadores relacionales que podemos utilizar en Python:

In [14]:
# "igual que"
1 == 1

True

In [15]:
# "diferente a"
"a" != "a"

False

In [16]:
# "mayor que"
10 > 5

True

In [17]:
# "menor que"
5 < 1

False

In [18]:
# "mayor o igual que"
30 >= 30

True

In [19]:
# "menor o igual que"
20 <= 10

False

Hay que tener cuidado y verificar que al hacer comparaciones los objetos implicados sean compatibles. Cuando los objetos no son *comparables* Python devolverá un `TypeError`, por ejemplo:

In [20]:
"a" > 10

TypeError: '>' not supported between instances of 'str' and 'int'

In [ ]:
[0,4] < (1,2)

También podemos *encadenar* comparaciones con instrucciones del tipo `a < b < c`, donde ese `<` puede ser cualquier operador relacional. Por ejemplo:

In [21]:
1 < 2 < 3

True

El `True` devuelto en la línea anterior implica que tanto `1 < 2` como `2 < 3` son condiciones que se cumplen. Veamos otros ejemplos:

In [22]:
10 > 10 > 3

False

In [23]:
3 == 3 >= 2

True

Este *encadenamiento* de operadores relacionales puede suplirse con el uso de operadores lógicos.

## Operadores lógicos

Los **operadores lógicos** nos sirven para realizar operaciones de lógica booleana entre valores de tipo `bool`. En Python podemos utilizar los operadores lógicos `and`, `or` y `not`. En la {numref}`op-log` se muestra la tabla de verdad para los valores lógicos.

```{list-table} Operadores lógicos
:header-rows: 1
:name: op-log

* - X
  - Y
  - X `and` Y
  - X `or` Y
  - `not` X
* - True
  - True
  - True
  - True
  - False
* - True
  - False
  - False 
  - True
  - False
* - False
  - True
  - False
  - True
  - True 
* - False
  - False
  - False 
  - False
  - True
```

Observemos los siguientes ejemplos:

In [24]:
True and True

True

In [25]:
True or False

True

In [26]:
not True

False

En un programa, la forma más habitual de uso de los operadores lógicos es encadenar múltiples comparaciones realizadas mediante los operadores relacionales. Debemos recordar que de una comparación resulta, invariablemente, un valor lógico `True` o `False`.

En el siguiente ejemplo se verifica si 1 es igual a 1 (verdadero) y si 2 es mayor a 1 (verdadero). Dado que ambas condiciones son verdaderas, entonces al encadenar ambas comparaciones con el operador `and`, resulta un valor lógico `True`, de acuerdo con la tabla de verdad mostrada anteriormente.

In [27]:
(1 == 1) and (2 > 1)

True

El operador lógico `or` podemos utilizarlo cuando una de ambas comparaciones debe ser cierta. Por ejemplo:

In [29]:
(0 != 0) or (10 > 20)

False

En la línea anterior, ambas comparaciones son falsas, entonces al encadenarlas con el operador `or` devuelven un valor lógico `False`. Si al menos una de las comparaciones es cierta, entonces se devuelve un `True`, tal como se muestra en el siguiente ejemplo:

In [30]:
(0 != 0) or (10 < 20)

True

 El uso habitual de los operadores lógicos es el encadenamiento de comparaciones, para tomar decisiones en el flujo del programa. Por ejemplo, si piensa en el caso de una línea de código que únicamente deberá ejecutarse en el caso de que se cumplan, sin excepción, dos condiciones, usando comparaciones en conjunto con el operador `and` y una estructura de bifurcación (algo que se revisará en secciones posteriores) se puede crear este tipo de programa.

## Expresiones

Las expresiones son combinaciones de constantes, variables, operadores, paréntesis y llamadas a funciones, que son evaluadas y producen un resultado. Son fundamentales porque constituyen la base para realizar cálculos o tomar decisiones en un lenguaje de programación. Podemos distinguir, generalmente, tres tipos de expresiones:

* Expresiones aritméticas
* Expresiones relacionales
* Expresiones lógicas

Las **expresiones aritméticas** utilizan únicamente operadores aritméticos combinados con cantidades numéricas (enteras o de coma flotante) y son utilizadas para evaluar fórmulas o expresiones algebraicas. Observa por ejemplo lo siguiente:

In [36]:
x = 10
y = 3*x**2 + 3*x - 2
print(y)

328


El valor que *guardamos* en la variable `y` es lo que resulta de evaluar la expresión $3x^2 + 3x - 2$ en $x=10$.

Es importante tener cuidado con la precedencia de los operadores al momento de crear una expresión aritmética, veamos un ejemplo:

In [40]:
a = 5
b = 3
c = 2
r1 = a / b + c
r2 = a / (b + c)
print("r1 =", r1)
print("r2 =", r2)

r1 = 3.666666666666667
r2 = 1.0


La diferencia en las expresiones anteriores, asignadas a `r1` y `r2`, radica en el uso del paréntesis. Mientras que para `r1` se está evaluando la expresión:

$$ r_1 = \frac{a}{b} + c $$

Para `r2` se evalúa:

$$ r_2 = \frac{a}{b+c} $$

Los paréntesis en la expresión asignada a `r2` indican que su contenido deberá evaluarse primero y posteriormente el resto de las operaciones aritméticas de acuerdo con la siguiente lista de precedencia de operadores (ordenada de mayor a menor prioridad):

* Paréntesis: `()` 
* Exponenciación: `**`.
* Multiplicación, división y módulo: `*`, `/`, `//`, `%`.
* Suma y resta: `+`, `-`.

Si se tiene una expresión con dos o más operadores con la misma prioridad, entonces se comenzará evaluando de izquierda a derecha. Por ejemplo:

In [47]:
3 / 5 * 2

1.2

En la expresión anterior, dado que los operadores `*` y `/` tienen la misma prioridad, entonces se evalúa primero el que está ubicado más a la izquierda (`/`), es decir:

$$ 3 / 5 * 2 = 0.6 * 2 = 1.2 $$

Para *forzar* que primero se evalúe el operador `*` entonces tendríamos que hacer uso de los paréntesis:

In [49]:
3 / (5 * 2)

0.3

Lo cual implicaría primero evaluar lo que está entre paréntesis y luego el resto, es decir:

$$ 3 / (5*2) = 3 / 10 = 0.3 $$

Las **expresiones lógicas** están formadas, típicamente, por operadores relacionales y lógicos, en conjunto con objetos de cualquier tipo de dato. Estas expresiones nos sirven para hacer comparaciones y combinarlas mediante los operadores lógicos, con la finalidad de *tomar decisiones* en el flujo de ejecución de un programa. Veamos una expresión de este tipo:

In [51]:
3 < 10 and 10 == 10

True

Observa que primero se hacen las comparaciones y luego se aplica el operador lógico `and`. Las expresiones lógicas pueden contener también expresiones aritméticas, por ejemplo:

In [54]:
1 + 2 + 3 > 10 and 5 + 3 > 7 

False

En la expresión anterior, tienen prioridad los operadores aritméticos sobre los relacionales, y estos sobre los lógicos. El proceso de evaluación sería como se muestra:

$$ 
1 + 2 + 3 > 10 \text{ and } 5 + 3 > 7 \\
\downarrow \\
6 > 10 \text{ and } 8 > 7 \\
\downarrow \\
\text{False} \text{ and } \text{True} \\
\downarrow \\
\text{False}
$$

Así pues, una lista *completa* de prioridad de operadores se muestra enseguida:

* Paréntesis: `()` 
* Exponenciación: `**`.
* Multiplicación, división y módulo: `*`, `/`, `//`, `%`.
* Suma y resta: `+`, `-`.
* Operadores relacionales: `==`, `!=`, `>`, `<`, `>=`, `<=`.
* Operadores lógicos: `not`, `and`, `or`.

## Ejemplos

<hr>
    
*Escribe un programa que calcule y muestre la raíz cuadrada de un número `n` dado.*

A continuación, se muestra una manera de resolver lo anterior:

In [82]:
n = 81
rc = (n)**(1/2)
print(f"La raíz cuadrada de {n} es {rc}")

La raíz cuadrada de 81 es 9.0


Observa que lo que estamos haciendo es evaluar la expresión:

$$ n^{1/2} $$

En el entendido de que es equivalente a la raíz cuadrada de un número. Si quisiéramos hacer nuestro programa un poco más interactivo podríamos utilizar la función `input`, la cual recibe por consola un valor en forma de una cadena de caracteres, veamos el ejemplo:

In [83]:
n = float( input("Ingresa un número: ") )
rc = (n)**(1/2)
print(f"La raíz cuadrada de {n} es {rc}")

Ingresa un número: 225
La raíz cuadrada de 225.0 es 15.0


En el código anterior la función `float` nos sirve para convertir la cadena de caracteres en un valor numérico, al cual se le pueda operar matemáticamente. Por ahora no te preocupes demasiado por la función `input`, en capítulos posteriores hablaremos con mayor detenimiento sobre esta.

<hr>

*La siguiente expresión nos permite aproximar el valor de $\pi$ hasta las millonésimas:*

$$
\pi \approx 3 + \frac{8}{60} + \frac{29}{60^2} + \frac{44}{60^3}
$$

*Escribe la expresión correspondiente en Python*

En Python podemos escribir dicha expresión como:

In [78]:
pi = 3 + (8/60) + (29/60**2) + (44/60**3)
pi

3.1415925925925925

Es importante tener en cuenta que en este caso los paréntesis no son realmente necesarios, observa:

In [79]:
pi = 3 + 8/60 + 29/60**2 + 44/60**3
pi

3.1415925925925925

Sin embargo el uso de estos suele aportar una mayor legibilidad a las expresiones matemáticas que estamos escribiendo. Hay que tener mucho cuidado con los paréntesis y la [precedencia de los operadores](https://docs.python.org/es/3/reference/expressions.html#operator-precedence), veamos el ejemplo siguiente:

In [80]:
pi = 3 + (8/60) + (29/60)**2 + (44/60)**3
pi

3.7613148148148148

Observa que la expresión colocada ya no corresponde con la aproximación de $\pi$ descrita con anterioridad, sino más bien correspondería a:

$$ 3 + \frac{8}{60} + \left(\frac{29}{60}\right)^2 + \left(\frac{44}{60}\right)^3 $$

<hr>

*Escribe un programa que evalúe la siguiente expresión para un valor dado de $x$*

$$ \frac{\sqrt{x}}{x^2+1} $$

In [70]:
x = 2 # valor de x
y = x**(1/2) / (x**2 + 1) # expresión
print(y) # mostramos el valor de la expresión evaluada

0.282842712474619


Observa que en el código anterior los paréntesis son necesarios, dado que si prescindimos por ejemplo del primer paréntesis:

In [71]:
x**1/2 / (x**2 + 1) 

0.2

Esto equivaldría a tener la expresión:

$$ \frac{\frac{x^1}{2}}{x^2 + 1} $$

Puesto que primero se evaluaría el exponente antes que el operador de división.

<hr>

*La energía cinética de una partícula está dada por:*

$$ K = \frac{1}{2} m v^2 $$

*Desarrolle un programa para evaluar la energía cinética para cuando $m=0.5 \,kg$ y $v=2.5 \, m/s^2$*

In [72]:
m = 0.5 
v = 2.5
K = 1/2 * m * v**2
print("K = ", K)

K =  1.5625


## Ejercicios

---
Crea un programa que dado un número `n` calcule y muestre la raíz cúbica de dicho número.

---
Realiza un programa que evalúe la siguiente expresión:

$$ 10^{0.375} - 30^{0.25} + \frac{2^2+10^2}{10^3}  $$

---
Crea un programa que evalúe y muestre el valor de la siguiente expresión:

$$ \sqrt{ 7 + \sqrt{6  +  \sqrt{5}} } $$

---
Crea un programa que evalúe y muestre el valor de la siguiente expresión:

$$ 5 + \frac{1}{1 - \frac{1}{5^{0.35}}} $$

---
Observa el siguiente código:

```python
a = 10
b = 20
c = 5
a = b + c
c = b + a
b = a + c
print(a, b, c)
```

¿Qué valores tendrán `a`, `b` y `c` al mostrarlos? Procura realizarlo primero de forma manual y enseguida comprueba tus resultados ejecutando el código anterior.

<hr>

El área de un círculo se determina utilizando la siguiente fórmula:

$$ A = \pi r^2 $$

Siendo $r$ el radio del círculo. Escribe un programa para calcular y mostrar el área de un círculo dado su radio.

<hr>

Para convertir de grados Fahrenheit a Celsius se puede utilizar la siguiente expresión:

$$ °C = \frac{5}{9}\left(°F - 32 \right) $$

Desarrolle un programa para calcular y mostrar la temperatura equivalente en grados Celsius, dada una temperatura en grados Fahrenheit.